rfd/xxxx-azure-node-join.md

```diff
@@ -0,0 +1,170 @@
+---
+authors: Andrew Burke (andrew.burke@goteleport.com)
+state: draft
+---
+
+# RFD X - Azure join method
+
+## Required Approvers
+Engineering: @jakule && @r0mant
+Product: @klizhentas && @xinding33
+Security: @reedloden
+
+## What
+
+Teleport nodes running on Azure virtual machines will be able to join a
+Teleport cluster without the need to share any secret token with the node.
+Instead, the node will present an
+[attested data document](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#attested-data)
+and an access token generated from the
+[instance metadata managed identity endpoint](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#managed-identity)
+to confirm the subscription the VM is running in. With a configured Azure join
+token on the auth server, the node will be allowed to join the cluster based
+on its Azure subscription.
+
+This is the Azure equivalent of EC2/IAM join as described in
+[RFD 41](https://github.com/gravitational/teleport/blob/master/rfd/0041-aws-node-join.md).
+
+## Why
+
+This will make provisioning nodes on Azure simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+In place of a join token, nodes will present a JWT access token and an attested
+data document to the auth server, both of which will be fetched on the nodes
+via instance metadata. The attested data request can include a user-specified
+nonce which will also be signed. Like the custom headers in IAM join, the nonce
+allows Teleport to issue a challenge (crypto random string of bytes) that the
+node must include in the signed request. We can use gRPC bidirectional streams
+to implement the following flow at a new `RegisterWithAzureToken` gRPC endpoint:
+
+1. The Node initiates a `RegisterWithAzureToken` request.
+2. The auth server enforces that TLS is used for transport.
+3. The auth server sends a base64 encoded 24 byte crypto-random challenge.
+  - We use 24 bytes instead of 32 because the nonce parameter is limited to 32
+    characters, and 24 bytes require 32 base64 characters.
+4. The node requests a) the attested data document using the challenge as the
+  nonce and b) an access token. Azure signs and returns both.
```

Comment:
I think it should be explained somewhere in this doc why both the attested data document and the access token are used. Perhaps a breakdown of each field that is used from those and each security measure that is in place.

My understanding is that the access token contains all the info about the VM we actually need, but it does not include a nonce, so we use the attested data document just for the nonce.

Is there any way to link the attested data document to the access token by VM ID or anything more specific than subscription ID? If not, I think any security benefit the attested data document provides could be bypassed by anyone with access to any VM in the subscription (even outside the allowed resource group). This weakens the effectiveness of the nonce since access tokens could still be reused by anyone able to generated attested data documents for the matching subscription. If you can link the access token to the same VM the attested data document is from somehow then I think this would be prevented.

^ I'm not saying this would be unacceptable either way, just that any benefit of the attested data document may be very limited. The access token by itself may be enough, I think it is basically equivalent to our oauth-based join methods.

Reply:
Your understanding is correct. The attested data does provide the VM ID, but there's no way to link that ID to the VM name and resource group without using the Azure API, which wouldn't work for Cloud.

Requiring the attested data at least has the benefit of ensuring that the VM is running, which EC2 join has to use the AWS API to do.

Reply:
> Requiring the attested data at least has the benefit of ensuring that the VM is running

It ensures that some VM in the subscription is running, but if you imagine an attacker controls a VM in a different resource group than the allowed one, they could generate the attested data on that VM. So in that case the only effective security measure is the access token. Which is why I was hoping the access token could be linked to it somehow

Reply:
Can this access token be used by Teleport to make a request to Azure API and fetch VM details somehow? If it has proper permissions, that is.

Reply:
https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/get?tabs=HTTP
https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/get?tabs=HTTP#security
I tested it, and it's possible! And the results include the VM ID, so we can link the access token and the attested data.